### PR TITLE
Code Improvements for Post-Reboot Script Calls

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
Code changes/additions to improve the chances of the post-reboot wait timeouts to succeed before executing our script calls. Also, a system log message for success or failure is left behind.

### CODE REVIEW

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop.

Even though the previous code already waited for NTP to be ready/synced, I explicitly added a conditional to wait for the "WAN connected" state just to be sure the router is indeed ready to network over the WAN. I also increased the wait timeout to 10 minutes as you had done, just in case there are edge cases where it takes a longer time for WAN to reach the **connected** state. If WAN is not ready within 10 minutes, something else must have gone wrong. Without recreating the actual scenario and investigating afterward, I cannot think of anything else to check at this point.

Please take a look at the PR changes when you get a chance.

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop.
> 
> Even though the previous code already waited for NTP to be ready/synced, I explicitly added a conditional to wait for the "WAN connected" state just to be sure the router is indeed ready to network over the WAN. I also increased the wait timeout to 10 minutes as you had done, just in case there are edge cases where it takes a longer time for WAN to reach the **connected** state. If WAN is not ready within 10 minutes, something else must have gone wrong. Without recreating the actual scenario and investigating afterward, I cannot think of anything else to check at this point.
> 
> Please take a look at the PR changes when you get a chance.


I Noticed you used the nvram values instead of doing an actual live connection test (such as to 8.8.8.8 or the ISP dns)
Was there a reason you opted to rely on the routers nvram values instead of just doing a live test for external connectivity?

More of a curiosity; it's not the original method I had in mind when I discussed doing a connection check in the last PR

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

As always your input is highly appreciated!!!

File: MerlinAU.sh [hunk 1]
```diff
@@ -1000,6 +1001,9 @@ _Init_Custom_Settings_Config_()
        sed -i "13 i FW_New_Update_Changelog_Approval=TBD" "$SETTINGSFILE"
        retCode=1
    fi
+   dos2unix "$SETTINGSFILE"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1004]: Good idea! Approved! While I appreciate this improvement, I don't think it was the cause of the problem

**Review by ExtremeFiretop — APPROVED** [review]

File: MerlinAU.sh [hunk 2]
```diff
@@ -6316,63 +6320,94 @@ Please manually update to version ${GRNct}${MinSupportedFirmwareVers}${NOct} or
...
       sleep $theWaitDelaySecs
       curWaitDelaySecs="$((curWaitDelaySecs + theWaitDelaySecs))"
    done
 
+   if [ "$curWaitDelaySecs" -lt "$maxWaitDelaySecs" ]
+   then Say "$logMsg was successful."
+   else Say "$logMsg [$maxWaitDelaySecs sec.] expired."
+   fi
+
+   sleep 30  ## Let's wait a bit & proceed ##
```
[the remark this answers is not in the document]
Comment (Martinski4GitHub) [reply MerlinAU.sh]: > Any thoughts on increasing this to 60?
>
>
I've just increased the final sleep delay before proceeding with the actual script function call to 60 sec. Originally, I saw no need to delay any further, but it doesn't hurt either. Of course, all this still needs to be tested.

**Comment by Martinski4GitHub** [comment]
> > @ExtremeFiretop.
> > Even though the previous code already waited for NTP to be ready/synced, I explicitly added a conditional to wait for the "WAN connected" state just to be sure the router is indeed ready to network over the WAN. I also increased the wait timeout to 10 minutes as you had done, just in case there are edge cases where it takes a longer time for WAN to reach the **connected** state. If WAN is not ready within 10 minutes, something else must have gone wrong. Without recreating the actual scenario and investigating afterward, I cannot think of anything else to check at this point.
> > Please take a look at the PR changes when you get a chance.
> 
>
> I noticed you used the nvram values instead of doing an actual live connection test (such as to 8.8.8.8 or the ISP dns) Was there a reason you opted to rely on the routers nvram values to determine true/external connectivity instead of just doing a live test for true/external connectivity?
> 
>
The main reason is that the NVRAM variables are updated as a result of the WAN events so when the WAN port reaches the "connected" state the NVRAM vars are updated accordingly to reflect its latest true state (this is in the C source code from RMerlin's GitHub repo. I saw this about 3-4 years ago and, AFAIK, that has not changed (I'll try to find the source code for it).

IMO, as long as NVRAM is mounted properly during reboot, relying on the NVRAM values is the way to go. No need to add more shell script code that basically does what the C source code does for us automatically. Of course, if the NVRAM interface is not working properly, all bets are off and we have a much bigger problem on our hands.

Since you were able to recreate the issue, try to test with the latest PR changes when you get a chance. If it happens, we need to have a better idea of what's going on, and specifically which area of the code to take a closer look at. Hopefully, the new syslog messages for success or failure will at least tell us whether the post-reboot wait timeout has something to do with it.

I'll take a look at your other messages & respond when I come back online later in the evening. I have an engineering meeting at 7:30 AM so I got to get some breakfast first & be ready for it.

Talk to you later, bud.

File: MerlinAU.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (ExtremeFiretop) [reply MerlinAU.sh]: @Martinski4GitHub 

I explained in more detail on the other PR why I think it may be helpful :)

**Comment by ExtremeFiretop** [comment]
> The main reason is that the NVRAM variables are updated as a result of the WAN events so when the WAN port reaches the "connected" state the NVRAM vars are updated accordingly to reflect its latest true state (this is in the C source code from RMerlin's GitHub repo. I saw this about 3-4 years ago and, AFAIK, that has not changed (I'll try to find the source code for it).
> 
> IMO, as long as NVRAM is mounted properly during reboot, relying on the NVRAM values is the way to go. No need to add more shell script code that basically does what the C source code does for us automatically. Of course, if the NVRAM interface is not working properly, all bets are off and we have a much bigger problem on our hands.
> 

My concern with nvram values is always a difference of values between models and configurations, but considering your just looping though the only 2 possible values; this is an acceptable solution and a similar one to what I used for RTRMON.

> Since you were able to recreate the issue, try to test with the latest PR changes when you get a chance. If it happens, we need to have a better idea of what's going on, and specifically which area of the code to take a closer look at. Hopefully, the new syslog messages for success or failure will at least tell us whether the post-reboot wait timeout has something to do with it.
> 

I also added logging to the services-start so we can see what happens there (a step up) when the issue is recreated! :)

> I'll take a look at your other messages & respond when I come back online later in the evening. I have an engineering meeting at 7:30 AM so I got to get some breakfast first & be ready for it.
> 
> Talk to you later, bud.

Sounds good! Enjoy buddy!

**Comment by ExtremeFiretop** [comment]
I just tested this version and still had the issue unfourtnantly; no email sent.
But if I manually ran: [ -f /jffs/scripts/MerlinAU.sh ] && sh /jffs/scripts/MerlinAU.sh postUpdateEmail &  #Added by MerlinAU#

It sent.... And I can confirm from the logs I added in services-start that it was triggered/called.

**Comment by ExtremeFiretop** [comment]
Hmm, this change seems to have introduced a bug? Or more like uncovered the bug I believe?

Let me test more and advise.

**Comment by Martinski4GitHub** [comment]
> > The main reason is that the NVRAM variables are updated as a result of the WAN events so when the WAN port reaches the "connected" state the NVRAM vars are updated accordingly to reflect its latest true state (this is in the C source code from RMerlin's GitHub repo. I saw this about 3-4 years ago and, AFAIK, that has not changed (I'll try to find the source code for it).
> > IMO, as long as NVRAM is mounted properly during reboot, relying on the NVRAM values is the way to go. No need to add more shell script code that basically does what the C source code does for us automatically. Of course, if the NVRAM interface is not working properly, all bets are off and we have a much bigger problem on our hands.
> 
> My concern with nvram values is always a difference of values between models and configurations, but considering your just looping though the only 2 possible values; this is an acceptable solution and a similar one to what I used for RTRMON. 
>
>
Are you saying that the values of the NVRAM vars for the "WAN state" are always different for different router models & configurations?

I'm not aware of that and have not seen RMerlin mention that there are differences in NVRAM values for the WAN state.
Do you get different values for your own routers?

>
> I'm assuming the value of state_t would be the expected value even if it was a dual-WAN setup for example?
> 
>
Yes, the code checks first for the "primary" WAN port & then its current state. There can be only one "primary" WAN port, regardless of a dual-WAN setup.

**Comment by Martinski4GitHub** [comment]
> I just tested this version and still had the issue unfortunately; no email sent. But if I manually ran: [ -f /jffs/scripts/MerlinAU.sh ] && sh /jffs/scripts/MerlinAU.sh postUpdateEmail & #Added by MerlinAU#
> 
> It sent.... And I can confirm from the logs I added in services-start that it was triggered/called.
>
What did the syslog file have at the time related to the post-reboot wait timeouts?
Did the script call ever reach the code to start the wait timeouts?
If yes, what was the result in the syslog?
If not, then something else might be failing before reaching the code.

**Comment by Martinski4GitHub** [comment]
> Hmm, this change seems to have introduced a bug? Or more like uncovered the bug I believe?
> 
>
"... this change ..." ???

**Comment by ExtremeFiretop** [comment]
> Are you saying that the values of the NVRAM vars for the "WAN state" are always different for different router models & configurations?

No I was talking "generally" why I avoid nvram values is due to the difference between configurations and hardware.
Thats why I said thats "always" my concern and in this case your just looping through 2 possible values so it's not a concern :)

**Comment by ExtremeFiretop** [comment]
> What did the syslog file have at the time related to the post-reboot wait timeouts? 

It didn't run at all!

>Did the script call ever reach the code to start the wait timeouts?

The services-start called everything inside; including our script, but our script is closing pre-maturely and not continuing to run in the background as we would expect :(

> If yes, what was the result in the syslog? If not, then something else might be failing before reaching the code.

Exactly! Now on to my old PR to explain this bit after I explain In detail more here.

**Comment by Martinski4GitHub** [comment]
> > What did the syslog file have at the time related to the post-reboot wait timeouts?
> 
> It didn't run at all!
> 
> > Did the script call ever reach the code to start the wait timeouts?
> 
> The services-start called everything inside; including our script, but our script is closing pre-maturely and not continuing to run in the background as we would expect :(
> 
> > If yes, what was the result in the syslog? If not, then something else might be failing before reaching the code.
> 
> Exactly! Now on to my old PR to explain this bit after I explain In detail more here.
>
OK, we have at least eliminated the post-reboot wait timeouts are a possible source of the problem.
This means something during the initialization phase is failing and terminates the execution of the script.

**Comment by ExtremeFiretop** [comment]
> > Hmm, this change seems to have introduced a bug? Or more like uncovered the bug I believe?
> 
> "... this change ..." ???

I realize this wasn't very descriptive, but at the time I commented it I had just noticed it; and didn't have a cause yet.
When I said "this change" i meant something in this PR. But now I think this is just unmasking part of the issue we have with the emails.

Post reboot after the upgrade I now have this settings file in the wrong directory:

image

The contents contains this:

image

I can confirm the services-start is being called and everything inside is being called and runs; and I can also confirm our script which is supposed to wait as a background job quits early and creates the above files.

If I add this change:

`readonly POST_UPDATE_EMAIL_SCRIPT_HOOK="[ -f $ScriptFilePath ] && sleep 120 && $POST_UPDATE_EMAIL_SCRIPT_JOB"`

To line 1476 the problem goes away, and the email sends as expected.

**Comment by Martinski4GitHub** [comment]
> > > Hmm, this change seems to have introduced a bug? Or more like uncovered the bug I believe?
> > 
> > 
> > "... this change ..." ???
> 
> I realize this wasn't very descriptive, but at the time I commented it I had just noticed it; and didn't have a cause yet. When I said "this change" i meant something in this PR. But now I think this is just unmasking part of the issue we have with the emails.
> 
> Post reboot after the upgrade I now have this settings file in the wrong directory:
> ...
> 
> The contents contains this:
> ....
> 
> I can confirm the services-start is being called and everything inside is being called and runs; and I can also confirm our script which is supposed to wait as a background job quits early and creates the above file.
> 
>
>
The "**/jffs/addons/custom_settings.txt**" file is a standard built-in file created for use by all add-ons installed. There's nothing unusual about its location, and it's not created by our script, but it can be modified by it.


> If I add this change:
> 
> `readonly POST_UPDATE_EMAIL_SCRIPT_HOOK="[ -f $ScriptFilePath ] && sleep 120 && $POST_UPDATE_EMAIL_SCRIPT_JOB"`
> 
> To line 1476 the problem goes away, and the email sends as expected.
>
>
This seems to indicate a timing problem during initialization or ....

P.S. I'm being called for dinner. Talk to you later...

**Comment by ExtremeFiretop** [comment]
> 
> The "**/jffs/addons/custom_settings.txt**" file is a standard built-in file created for use by all add-ons installed. There's nothing unusual about its location, and it's not created by our script, but it can be modified by it.
> 

Oh? Maybe your right, and I'm confused. This is news to me... It has the exact same name as our settings file. And I can't find it on any of my other 2 routers.

> > If I add this change:
> > `readonly POST_UPDATE_EMAIL_SCRIPT_HOOK="[ -f $ScriptFilePath ] && sleep 120 && $POST_UPDATE_EMAIL_SCRIPT_JOB"`
> > To line 1476 the problem goes away, and the email sends as expected.
> 
> This seems to indicate a timing problem during initialization or ....
> 

Agreed!

> P.S. I'm being called for dinner. Talk to you later...

Ttyl! I'll be around testing things. :)

**Comment by ExtremeFiretop** [comment]
Ah. It seems to be related with the Web add-ons? Is that correct?
Maybe that's why it's not on my other 2 routers and I'm not familiar in seeing it lol.

**Comment by Martinski4GitHub** [comment]
> Ah. It seems to be related with the Web add-ons? Is that correct? Maybe that's why it's not on my other 2 routers and I'm not familiar in seeing it lol.
>
>
Correct. The "**/jffs/addons/custom_settings.txt**" file is available for all add-ons to use but it's not a requirement. It's used mostly by add-ons that include a webGUI page to keep track of "local vs server" versions and as a mechanism to transfer small amounts of data between the HTML/JavaScript code and the shell script.

If a router doesn't have the file, it usually means there are no add-ons with a webGUI page installed.

**Pull request merged**